### ISSUE TITLE
HeaderSearch crashes with Microsoft search rules

### ISSUE THREAD

**Issue by mikerice1969** [llvm/llvm-project#54426]
To reproduce:
tar zxvf ../repro.tar.gz
clang -cc1 -E -fms-compatibility -o - -x c -Ishared -Ishared/driver v.c

Started failing after:
https://reviews.llvm.org/D119721/
https://github.com/llvm/llvm-project/commit/17c9fcd6f6fc0926c1096242a2ddced9b523decd

In file included from v.c:1:
In file included from shared/version/v.h:1:
In file included from shared/diag/d.h:1:
shared/syntax/t.h:1:10: warning: #include resolved using non-portable Microsoft search rules as: shared/diag/mc.h [-Wmicrosoft-include]
#include "mc.h"
         ^
In file included from v.c:1:
In file included from shared/version/v.h:1:
In file included from shared/diag/d.h:1:
In file included from shared/syntax/t.h:1:
shared/diag/mc.h:1:9: warning: including diag/mc.h [-W#pragma-messages]
#pragma message "including diag/mc.h"
        ^
clang: ws/llvm-project/clang/include/clang/Lex/HeaderSearch.h:196: clang::detail::Qualified<IsConst, clang::DirectoryLookup>& clang::detail::SearchDirIteratorImpl<IsConst>::operator*() const [with bool IsConst = true; clang::detail::Qualified<IsConst, clang::DirectoryLookup> = const clang::DirectoryLookup]: Assertion `*this && "Invalid iterator."' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang -cc1 -E -fms-compatibility -o - -x c -Ishared -Ishared/driver v.c
 #0 0x00007f60ea9ff261 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) ws/llvm-project/llvm/lib/Support/Unix/Signals.inc:565:22
 #1 0x00007f60ea9ff318 PrintStackTraceSignalHandler(void*) ws/llvm-project/llvm/lib/Support/Unix/Signals.inc:632:1
 #2 0x00007f60ea9fd2d4 llvm::sys::RunSignalHandlers() ws/llvm-project/llvm/lib/Support/Signals.cpp:94:20
 #3 0x00007f60ea9feca9 SignalHandler(int) ws/llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007f60e9280dd0 __restore_rt sigaction.c:0:0
 #5 0x00007f60e493370f raise (/lib64/libc.so.6+0x3770f)
 #6 0x00007f60e491db25 abort (/lib64/libc.so.6+0x21b25)
 #7 0x00007f60e491d9f9 _nl_load_domain.cold.0 loadmsgcat.c:0:0
 #8 0x00007f60e492bcc6 .annobin___GI___assert_fail.end assert.c:0:0
 #9 0x00007f60dd453fe1 clang::detail::SearchDirIteratorImpl<true>::operator*() const ws/llvm-project/clang/include/clang/Lex/HeaderSearch.h:196:5
#10 0x00007f60dd44ebd4 llvm::iterator_facade_base<clang::detail::SearchDirIteratorImpl<true>, std::forward_iterator_tag, clang::DirectoryLookup const, long, clang::DirectoryLookup const*, clang::DirectoryLookup const&>::operator->() const ws/llvm-project/llvm/include/llvm/ADT/iterator.h:206:59
#11 0x00007f60dd4409b6 clang::HeaderSearch::LookupFile(llvm::StringRef, clang::SourceLocation, bool, clang::detail::SearchDirIteratorImpl<true>, clang::detail::SearchDirIteratorImpl<true>*, llvm::ArrayRef<std::pair<clang::FileEntry const*, clang::DirectoryEntry const*> >, llvm::SmallVectorImpl<char>*, llvm::SmallVectorImpl<char>*, clang::Module*, clang::ModuleMap::KnownHeader*, bool*, bool*, bool, bool) ws/llvm-project/clang/lib/Lex/HeaderSearch.cpp:999:37
#12 0x00007f60dd4c5caf clang::Preprocessor::LookupFile(clang::SourceLocation, llvm::StringRef, bool, clang::detail::SearchDirIteratorImpl<true>, clang::FileEntry const*, clang::detail::SearchDirIteratorImpl<true>*, llvm::SmallVectorImpl<char>*, llvm::SmallVectorImpl<char>*, clang::ModuleMap::KnownHeader*, bool*, bool*, bool) ws/llvm-project/clang/lib/Lex/PPDirectives.cpp:903:53
#13 0x00007f60dd4c959d clang::Preprocessor::LookupHeaderIncludeOrImport(clang::detail::SearchDirIteratorImpl<true>*, llvm::StringRef&, clang::SourceLocation, clang::CharSourceRange, clang::Token const&, bool&, bool, bool&, clang::detail::SearchDirIteratorImpl<true>, clang::FileEntry const*, llvm::StringRef&, llvm::SmallVectorImpl<char>&, llvm::SmallVectorImpl<char>&, clang::ModuleMap::KnownHeader&, bool) ws/llvm-project/clang/lib/Lex/PPDirectives.cpp:1879:53
#14 0x00007f60dd4ca455 clang::Preprocessor::HandleHeaderIncludeOrImport(clang::SourceLocation, clang::Token&, clang::Token&, clang::SourceLocation, clang::detail::SearchDirIteratorImpl<true>, clang::FileEntry const*) ws/llvm-project/clang/lib/Lex/PPDirectives.cpp:2049:74
#15 0x00007f60dd4c92b5 clang::Preprocessor::HandleIncludeDirective(clang::SourceLocation, clang::Token&, clang::detail::SearchDirIteratorImpl<true>, clang::FileEntry const*) ws/llvm-project/clang/lib/Lex/PPDirectives.cpp:1844:79
#16 0x00007f60dd4c67d8 clang::Preprocessor::HandleDirective(clang::Token&) ws/llvm-project/clang/lib/Lex/PPDirectives.cpp:1106:36
#17 0x00007f60dd477eec clang::Lexer::LexTokenInternal(clang::Token&, bool) ws/llvm-project/clang/lib/Lex/Lexer.cpp:4093:7
#18 0x00007f60dd4753f0 clang::Lexer::Lex(clang::Token&) ws/llvm-project/clang/lib/Lex/Lexer.cpp:3307:40
#19 0x00007f60dd51a996 clang::Preprocessor::Lex(clang::Token&) ws/llvm-project/clang/lib/Lex/Preprocessor.cpp:898:36
#20 0x00007f60e65dbb07 clang::DoPrintPreprocessedInput(clang::Preprocessor&, llvm::raw_ostream*, clang::PreprocessorOutputOptions const&) ws/llvm-project/clang/lib/Frontend/PrintPreprocessedOutput.cpp:1006:15
#21 0x00007f60e65a35ca clang::PrintPreprocessedAction::ExecuteAction() ws/llvm-project/clang/lib/Frontend/FrontendActions.cpp:1050:27
#22 0x00007f60e65924bf clang::FrontendAction::Execute() ws/llvm-project/clang/lib/Frontend/FrontendAction.cpp:980:38
#23 0x00007f60e6455ed0 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) ws/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1036:42
#24 0x00007f60e5edb0ed clang::ExecuteCompilerInvocation(clang::CompilerInstance*) ws/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:263:38
#25 0x000000000045031a cc1_main(llvm::ArrayRef<char const*>, char const*, void*) ws/llvm-project/clang/tools/driver/cc1_main.cpp:248:40
#26 0x0000000000444e7f ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) ws/llvm-project/clang/tools/driver/driver.cpp:317:20
#27 0x00000000004454f0 main ws/llvm-project/clang/tools/driver/driver.cpp:388:26
#28 0x00007f60e491f6a3 __libc_start_main (/lib64/libc.so.6+0x236a3)
#29 0x00000000004437ee _start (/localdisk2/mprice1/community6/llvm-project/build/bin/clang-14+0x4437ee)

**Comment by mikerice1969** [comment]
@jansvoboda11  could you take a look?

In the reproducer there are two mc.h files in different directories.  The old (presumably working) behavior the diag/mc.h was picked up first, then later the driver/mc.h.  

Now the lookup of mc.h ends up with a bad iterator (HD is nullptr).

**Comment by jansvoboda11** [comment]
Thanks for reporting this, I'm happy to take a look. Can you attach the reproducer please?

**Comment by mikerice1969** [comment]
Sorry, here it is. [repro.tar.gz](https://github.com/llvm/llvm-project/files/8305413/repro.tar.gz)

**Comment by erichkeane** [comment]
@jansvoboda11 : Were you able to take a look at Mike's reproducer?

**Comment by jansvoboda11** [comment]
Yes, the reproducer was helpful. I have a fix up for review here: https://reviews.llvm.org/D122237.

**Comment by jansvoboda11** [comment]
Can you please confirm 59dadd178b0b fixes the crash you encountered?
